### PR TITLE
v3_1_0: Move k8s components to selfhosted pods

### PR DESCRIPTION
**Pull request by corest**
Towards: https://github.com/giantswarm/giantswarm/issues/2374

### CODE REVIEW

File: v_3_1_0/master_template.go [hunk 1]
```diff
@@ -1768,6 +1768,211 @@ write_files:
...
+    metadata:
+      name: k8s-api-server
+      namespace: kube-system
+    spec:
+      hostNetwork: true
+      containers:
+      - name: k8s-api-server
+        image: quay.io/giantswarm/hyperkube:v1.9.2
+        env:
+        - name: HOST_IP
```
Comment (corest) [inline v_3_1_0/master_template.go]: Difference from systemd unit, where we used `/etc/network-environment` for getting env variable with host ip address.

File: v_3_1_0/master_template.go [hunk 2]
```diff
@@ -1768,6 +1768,211 @@ write_files:
...
+        - --audit-log-maxbackup=30
+        - --audit-log-maxsize=100
+        - --audit-policy-file=/etc/kubernetes/manifests/audit-policy.yml
+        - --experimental-encryption-provider-config=/etc/kubernetes/encryption/k8s-encryption-config.yaml
+        resources:
+          requests:
+            cpu: 300m
+        livenessProbe:
+          tcpSocket:
+            port: {{.Cluster.Kubernetes.API.SecurePort}}
```
Comment (corest) [inline v_3_1_0/master_template.go]: Here is the tricky thing. 
I tried httpGet with `HTTPS` scheme, which fails with code `401`, which is actually healthy status code from the health perspective. 
TCP check seems enough for API liveness probe.

**Comment by corest** [comment]
Tested on gauss, got working guest cluster with self-hosted k8s components:
```
calico-kube-controllers-7c57ddf866-6425z                            1/1       Running   0          2m
calico-node-6s8fg                                                   2/2       Running   0          2m
calico-node-qmjnq                                                   2/2       Running   0          2m
coredns-84cd68c555-cnkl6                                            1/1       Running   0          2m
coredns-84cd68c555-jzwkv                                            1/1       Running   0          2m
default-http-backend-6d65799844-5nj8x                               1/1       Running   0          2m
default-http-backend-6d65799844-752zz                               1/1       Running   0          2m
k8s-api-server-ip-10-1-2-64.eu-central-1.compute.internal           1/1       Running   0          2m
k8s-controller-manager-ip-10-1-2-64.eu-central-1.compute.internal   1/1       Running   0          1m
k8s-scheduler-ip-10-1-2-64.eu-central-1.compute.internal            1/1       Running   0          1m
kube-proxy-6s66b                                                    1/1       Running   0          2m
kube-proxy-zm6nx                                                    1/1       Running   0          2m
kube-state-metrics-6b6bbbb948-r7m9t                                 1/1       Running   0          2m
nginx-ingress-controller-6f749fb46d-tps9w                           1/1       Running   0          2m
node-exporter-6rsnf                                                 1/1       Running   0          2m
node-exporter-7pd5c                                                 1/1       Running   0          2m
```

File: v_3_1_0/master_template.go [hunk 3]
```diff
@@ -2114,155 +2321,14 @@ coreos:
...
   - name: k8s-addons.service
     enable: true
     command: start
     content: |
       [Unit]
       Description=Kubernetes Addons
-      Wants=k8s-api-server.service
-      After=k8s-api-server.service
+      Wants=k8s-kubelet.service
+      After=k8s-kubelet.service
```
Comment (puja108) [inline v_3_1_0/master_template.go]: I guess this works without race condition even when it takes a while for kubelet to start api server?

Comment (corest) [reply v_3_1_0/master_template.go]: `k8s-addons` uses those checks in systemd only for proper order, real dependency is inside the script. So nothing happens before cluster is healthy

File: v_3_1_0/master_template.go (hunk 1 above)
Comment (r7vme) [reply v_3_1_0/master_template.go]: ack.

File: v_3_1_0/master_template.go (hunk 2 above)
Comment (r7vme) [reply v_3_1_0/master_template.go]: even `/healthz` gives you 401?

i see that kubeadm uses https https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/controlplane/manifests.go#L84

Comment (corest) [reply v_3_1_0/master_template.go]: I tested that path. Gives me `401`  as well

**Pull request merged**